[core][cgraph] Remove asyncio_max_queue_size and use max_buffered_results for async

dayshah opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This removes the asyncio_max_queue_size, and makes compiled graphs that use asyncio use the max_buffered_results paramter
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
Closes https://github.com/ray-project/ray/issues/49051
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

dayshah commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1968,6 +1957,13 @@ def _cache_execution_results(
+                f"buffered results is {self._max_buffered_results}; call "
+                "ray.get() on previous CompiledDAGRefs to free them up "
+                "from buffer."
+            )
```
cache execution results is used by both async and standard, so if we put the check here it will do it for both

kevin85421 reviewed (commented):
`asyncio_max_queue_size` and `max_buffered_results` have different definitions. 

* `asyncio_max_queue_size` represents the maximum number of "inputs" that have not yet been written into the DAG.

* `max_buffered_results` is the number of execution "results" that have already been retrieved by the driver but have not yet been consumed by the users.

Hence, removing `asyncio_max_queue_size` is a breaking change. It's fine to have a breaking change because the RayCG API is currently in alpha status, but we should have a clear plan for handling the gap (e.g., using `max_inflight_executions` to prevent users from sending unlimited inputs to the asyncio queue).

Would you mind writing more details in the PR description and adding tests if needed?

dayshah commented:
> @kevin85421 @stephanie-wang Correct me if wrong, but I thought we want to collapse both `max_buffered_results` and `asyncio_max_queue_size` into `max_inflight_executions`?

ya that makes sense actually because getting the results actually removes from the result_buffer and all that max_buffered_results controls is the result buffer size which can only go up to max_inflight_executions anyways, so will remove both and only keep max_inflight_executions.

dayshah commented:
> `asyncio_max_queue_size` and `max_buffered_results` have different definitions.
> 
> * `asyncio_max_queue_size` represents the maximum number of "inputs" that have not yet been written into the DAG.
> * `max_buffered_results` is the number of execution "results" that have already been retrieved by the driver but have not yet been consumed by the users.
> 
> Hence, removing `asyncio_max_queue_size` is a breaking change. It's fine to have a breaking change because the RayCG API is currently in alpha status, but we should have a clear plan for handling the gap (e.g., using `max_inflight_executions` to prevent users from sending unlimited inputs to the asyncio queue).
> 
> Would you mind writing more details in the PR description and adding tests if needed?

updated with description, I think the test that exists currently for max_inflight_executions seems ok

kevin85421 commented:
> @kevin85421 @stephanie-wang Correct me if wrong, but I thought we want to collapse both `max_buffered_results` and `asyncio_max_queue_size` into `max_inflight_executions`?

As I recall, this was not the conclusion, but I think removing both and using only `max_inflight_executions` would be better. The reason is that `max_buffered_results` is hard for users to understand, and it depends on the speed of the compiled graphs. Additionally, it is difficult for users to determine when to call ray.get because it depends on the processing speed of the DAG.

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -2068,22 +2046,15 @@ def _execute_until(
             # them separately to enable individual retrieval
             self._cache_execution_results(
-                self._max_finished_execution_index,
+                self._max_finished_execution_index + 1,
                 self._dag_output_fetcher.read(timeout),
             )
+            self.increment_max_finished_execution_index()
```
Why is this change?

dayshah commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -2068,22 +2046,15 @@ def _execute_until(
+                self._max_finished_execution_index + 1,
                 self._dag_output_fetcher.read(timeout),
             )
+            self.increment_max_finished_execution_index()
```
I actually made this change before when making max_buffered_results apply to async, but I think it still makes sense to make the logic consistent with async which only increments after caching and reading. In the old situation `max_finished_execution_index` would still be incremented if the read timed out
https://github.com/ray-project/ray/blob/a1c6c712fe021e3248a8ead683fa553b28806006/python/ray/experimental/compiled_dag_ref.py#L209-L210

ruisearch42 replied:
```
            result = self._dag_output_fetcher.read(timeout)
            self.increment_max_finished_execution_index()
            self._cache_execution_results(
                self._max_finished_execution_index,
                result,
            )
```
This is slightly cleaner?

BTW, can we remove the `increment_max_finished_execution_index()` method? We should not need it.

dayshah replied:
ya, that's cleaner I'll do that and update async to have the same order, and I'll remove the function

Will just move things related to this to a separate pr to keep this removal clean

dayshah replied:
just did it in this pr since i'm changing other execution_index counting things

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -2068,22 +2046,15 @@ def _execute_until(
                 timeout = ctx.get_timeout
 
         while self._max_finished_execution_index < execution_index:
-            if len(self._result_buffer) >= self._max_buffered_results:
-                raise ValueError(
-                    "Too many buffered results: the allowed max count for "
-                    f"buffered results is {self._max_buffered_results}; call "
-                    "ray.get() on previous CompiledDAGRefs to free them up "
-                    "from buffer."
-                )
```
I think we should not remove this check, but use `self._max_inflight_executions` as the threshold.

And here the "inflight" has a slight different interpretation: inflight meaning buffered but not retrieved. We should throw `RayCgraphCapacityExceeded` as well.

dayshah replied:
Ya this would actually cover the case I was mentioning yesterday where taking out result_buffer could get very big even with small inflight_executions if you do ray.get in a different order than dag.execute

So I'll add this back and also update the comment docs to clarify that max_inflight_executions counts an execution as until you do ray.get

Another thing is that I also have to update async to have this check in its codepath too

dayshah replied:
done

ruisearch42 reviewed (commented):
Otherwise LGTM

dayshah commented:
adjusted execution_index counting, since before the counts would be off by one, it still worked since we only checked > not >= which it should've been. Also added tests for both exceptions.

dayshah commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,41 @@ async def main():
+        ref3 = compiled_dag.execute(1)
+    # to show variables are being used and avoid destruction since
+    # CompiledDagRef __del__ will release buffers and
+    # increment _max_finished_execution_index
```
this seems prone to issues, ex. if destruction is not in the same order execute was called
https://github.com/ray-project/ray/blob/ea4e3152a3fdc6c097a804258d1338066b56ef63/python/ray/experimental/compiled_dag_ref.py#L103-L104
this is also inconsistent behavior vs. async

ruisearch42 replied:
Yeah that looks problematic. cc @kevin85421 who reviewed the skip deserialization PR.

One idea is to deserialize the prior indexes and save the results to buffer, and only skip deserialization for the exact index:
A ref goes out of scope => first check if result in buffer, if so delete, otherwise do above.

dayshah replied:
Ya that sounds ok, there's the overhead of needing to get the prior indices though. But not really any way to get around that. One thing is that python might not call __del__ in order so we may be forcing deserialization for refs that are also being deleted but not sure if we can do anything about that.

Should I open an issue to track, is it a beta blocker or just do this in a follow-up pr, no need for issue creation?

ruisearch42 replied:
I think we should ensure correctness first. But open an issue for optimization. We can leave the optimization out of beta for now.

ruisearch42 replied:
Looks like this one is not addressed yet?

dayshah replied:
opened https://github.com/ray-project/ray/pull/49781

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1934,16 +1906,26 @@ def raise_if_too_many_inflight_requests(self):
+    def raise_if_result_buffer_at_capacity(self):
+        if len(self._result_buffer) >= self._max_inflight_executions:
+            raise ray.exceptions.RayCgraphCapacityExceeded(
+                f"{len(self._result_buffer)} buffered results: the allowed max count "
```
mention something like "we cannot execute more" so that the error message makes sense when the two values are equal

ruisearch42 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1934,16 +1906,26 @@ def raise_if_too_many_inflight_requests(self):
-                f"There are {num_in_flight_requests} in-flight requests which "
-                "is more than specified _max_inflight_executions of the dag: "
+                f"{num_in_flight_requests} in-flight requests: "
+                "the specified _max_inflight_executions of the dag is "
```
```suggestion
                "the specified _max_inflight_executions of the DAG is "
```

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1934,16 +1906,26 @@ def raise_if_too_many_inflight_requests(self):
         num_in_flight_requests = (
```
It’s better to always raise an exception when users call `execute` rather than when they call `ray.get` in some cases.

Maybe we should change the definition of `num_in_flight_requests` in this function to accurately track the number of in-flight requests in the DAG.

dayshah replied:
done, adding len(result_buffer) to this and just checking here on execute

ruisearch42 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,41 @@ async def main():
+    ref1 = compiled_dag.execute(1)
+    ref2 = compiled_dag.execute(1)
+    with pytest.raises(
+        ray.exceptions.RayCgraphCapacityExceeded, match=(r"2 in-flight requests: ")
```
let's spell out the exact error message so the test becomes self-explanatory

ruisearch42 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,41 @@ async def main():
+    ref3 = compiled_dag.execute(3)
+    ref4 = compiled_dag.execute(4)
+    with pytest.raises(
+        ray.exceptions.RayCgraphCapacityExceeded, match=(r"2 buffered results: ")
```
ditto: spell out full exception

ruisearch42 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,41 @@ async def main():
     loop.run_until_complete(main())
 
 
+def test_inflight_requests_exceeds_max_inflight_requests(ray_start_regular):
```
`test_inflight_requests_exceed_capacity` for short

ruisearch42 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,41 @@ async def main():
+    (ref1, ref2, ref3)
+
+
+def test_result_buffer_exceeds_max_inflight_requests(ray_start_regular):
```
`test_result_buffer_exceeds_capacity` for short

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -1933,14 +1905,13 @@ def run(self):
-                f"{self._max_inflight_executions}. Retrieve the output using "
-                "ray.get before submitting more requests or increase "
-                "`max_inflight_executions`. "
+                f"You cannot execute more than {self._max_inflight_executions} "
```
Maybe change to "The compiled DAG cannot execute more than ..." instead?

dayshah replied:
the user is the one doing the executing and calling execute though. Can do something like 
`you can't execute the compiled graph more than x times`, but I feel like it's pretty obvious it's a compiled graph execution because the error will point to the compiled_graph.execute

ruisearch42 replied:
+1 to what Kai-Hsun said. I also feel it's a bit weird to say "you ..." , can we avoid that? Using `There are ... ` or `The compiled DAG/graph` sounds better.

dayshah replied:
update to 
```
                "The compiled graph can't have more than "
                f"{self._max_inflight_executions} in-flight executions, and you "
                f"currently have {num_inflight_executions} in-flight executions. "
                "Retrieve an output using ray.get before submitting more requests or "
                "increase `_max_inflight_executions`. "
                "`dag.experimental_compile(_max_inflight_executions=...)`"
```

kevin85421 commented on python/ray/experimental/compiled_dag_ref.py:
```diff
@@ -205,9 +205,10 @@ def __await__(self):
         self._fut = None
 
         if not self._dag._has_execution_results(self._execution_index):
+            self._dag.raise_if_result_buffer_at_capacity()
```
what's this?

dayshah replied:
removed

kevin85421 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -2252,6 +2174,52 @@ async def main():
     loop.run_until_complete(main())
 
 
+def test_inflight_requests_exceed_capacity(ray_start_regular):
```
test for both `execute` and `execute_async`?

dayshah replied:
added

kevin85421 reviewed (approved):
Others LGTM

ruisearch42 reviewed (commented):
Code has a conflict with master.
